File: benchmarks/json_parse/json_parse.js
const SRC = '{"users":[{"id":1,"name":"Alice","active":true,"scores":[85,92,78]},{"id":2,"name":"Bob","active":false,"scores":[91,88,95]},{"id":3,"name":"Charlie","active":true,"scores":[76,84,90]}],"metadata":{"version":2,"generated":false,"tags":["benchmark","json","test"],"config":{"timeout":30,"retries":3}}}';

function Parser(src) {
    this.src = src;
    this.pos = 0;
}

Parser.prototype.peek = function() { return this.src.charCodeAt(this.pos); };
Parser.prototype.advance = function() { this.pos++; };
Parser.prototype.eof = function() { return this.pos >= this.src.length; };

Parser.prototype.isDigit = function(c) { return c >= 48 && c <= 57; };
Parser.prototype.isSpace = function(c) { return c === 32 || c === 9 || c === 10 || c === 13; };

Parser.prototype.skipWS = function() {
    while (!this.eof() && this.isSpace(this.peek())) this.advance();
};

Parser.prototype.parseString = function() {
    this.advance();
    var n = 0;
    while (this.peek() !== 34) { this.advance(); n++; }
    this.advance();
    return n;
};

Parser.prototype.parseNumber = function() {
    var sign = 1;
    if (this.peek() === 45) { this.advance(); sign = -1; }
    var val = 0;
    while (!this.eof() && this.isDigit(this.peek())) {
        val = val * 10 + (this.peek() - 48);
        this.advance();
    }
    return sign * val;
};

Parser.prototype.parseValue = function() {
    this.skipWS();
    if (this.eof()) return 0;
    var c = this.peek();
    if (c === 34) return this.parseString();
    if (c === 116) { this.advance(); this.advance(); this.advance(); this.advance(); return 1; }
    if (c === 102) { this.advance(); this.advance(); this.advance(); this.advance(); this.advance(); return 0; }
    if (c === 110) { this.advance(); this.advance(); this.advance(); this.advance(); return 0; }
    if (c === 45 || this.isDigit(c)) return this.parseNumber();
    if (c === 91) return this.parseArray();
    if (c === 123) return this.parseObject();
    return 0;
};

Parser.prototype.parseArray = function() {
    this.advance(); this.skipWS();
    if (this.peek() === 93) { this.advance(); return 0; }
    var sum = 0;
    while (true) {
        sum += this.parseValue();
        this.skipWS();
        if (this.peek() === 44) this.advance();
        else break;
    }
    this.advance();
    return sum;
};

Parser.prototype.parseObject = function() {
    this.advance(); this.skipWS();
    if (this.peek() === 125) { this.advance(); return 0; }
    var sum = 0;
    while (true) {
        this.skipWS();
        this.parseString();
        this.skipWS();
        this.advance();
        sum += this.parseValue();
        this.skipWS();
        if (this.peek() === 44) this.advance();
        else break;
    }
    this.advance();
    return sum;
};

var parser = new Parser(SRC);
var total = 0;
for (var i = 0; i < 100000; i++) {
    parser.pos = 0;
    total += parser.parseValue();
}
console.log(total);

File: benchmarks/json_parse/json_parse.c
#include <stdio.h>
#include <string.h>

static const char *src = "{\"users\":[{\"id\":1,\"name\":\"Alice\",\"active\":true,\"scores\":[85,92,78]},{\"id\":2,\"name\":\"Bob\",\"active\":false,\"scores\":[91,88,95]},{\"id\":3,\"name\":\"Charlie\",\"active\":true,\"scores\":[76,84,90]}],\"metadata\":{\"version\":2,\"generated\":false,\"tags\":[\"benchmark\",\"json\",\"test\"],\"config\":{\"timeout\":30,\"retries\":3}}}";

typedef struct { const char *restrict src; int pos; int len; } Parser;

static int peek(Parser *p) { return p->src[p->pos]; }
static void advance(Parser *p) { p->pos++; }
static int eof(Parser *p) { return p->pos >= p->len; }
static int is_digit(int c) { return c >= '0' && c <= '9'; }
static int is_space(int c) { return c == ' ' || c == '\t' || c == '\n' || c == '\r'; }

static void skip_ws(Parser *p) {
    while (!eof(p) && is_space(peek(p))) advance(p);
}

static long long parse_string(Parser *p) {
    advance(p);
    long long len = 0;
    while (peek(p) != '"') {
        advance(p);
        len++;
    }
    advance(p);
    return len;
}

static long long parse_number(Parser *p) {
    long long sign = 1;
    if (peek(p) == '-') { advance(p); sign = -1; }
    long long val = 0;
    while (!eof(p) && is_digit(peek(p))) {
        val = val * 10 + (peek(p) - '0');
        advance(p);
    }
    return sign * val;
}

static long long parse_value(Parser *p);

static long long parse_array(Parser *p) {
    advance(p); skip_ws(p);
    if (peek(p) == ']') { advance(p); return 0; }
    long long sum = 0;
    for (;;) {
        sum += parse_value(p);
        skip_ws(p);
        if (peek(p) == ',') advance(p);
        else break;
    }
    advance(p);
    return sum;
}

static long long parse_object(Parser *p) {
    advance(p); skip_ws(p);
    if (peek(p) == '}') { advance(p); return 0; }
    long long sum = 0;
    for (;;) {
        skip_ws(p);
        parse_string(p);  /* discard key */
        skip_ws(p);
        advance(p);  /* ':' */
        sum += parse_value(p);
        skip_ws(p);
        if (peek(p) == ',') advance(p);
        else break;
    }
    advance(p);
    return sum;
}

static long long parse_value(Parser *p) {
    skip_ws(p);
    if (eof(p)) return 0;
    int c = peek(p);
    if (c == '"') return parse_string(p);
    if (c == 't') { advance(p); advance(p); advance(p); advance(p); return 1; }
    if (c == 'f') { advance(p); advance(p); advance(p); advance(p); advance(p); return 0; }
    if (c == 'n') { advance(p); advance(p); advance(p); advance(p); return 0; }
    if (c == '-' || is_digit(c)) return parse_number(p);
    if (c == '[') return parse_array(p);
    if (c == '{') return parse_object(p);
    return 0;
}

int main(void) {
    Parser parser = { .src = src, .pos = 0, .len = (int)strlen(src) };
    long long total = 0;
    for (int i = 0; i < 100000; i++) {
        parser.pos = 0;
        total += parse_value(&parser);
    }
    printf("%lld\n", total);
    return 0;
}
